### ISSUE TITLE
RunPostOrder and FetchTensorData methods adds overhead and prevents complete pipelining of two executions

### ISSUE THREAD

**Issue by aws-rhsoln** [pytorch/xla#3434]
## 🐛 Bug

When training model with multiple workers, two `ExecuteComputation` are not completely pipelined. The two `ExecuteComputation` are separated by execution of two methods RunPostOrder and FetchTensorData. These two methods end up taking around 5-7% of the total execution. Here is the profile from the MNIST script ran on GPU instance with 8 workers.
![Screen Shot 2022-03-21 at 1 21 12 PM](https://user-images.githubusercontent.com/73188662/159359570-905a1966-186e-4318-a4d3-e2d4c9752722.png)

Here the highlighted area corresponds to the time spent in [RunPostOrder](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1272) and [FetchTensorData](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1312) methods. Each ExecuteComputation takes around 1.8ms.

I tried increasing the batchsize so that more time is spent on ExecutionComputation, and see if I can overlap the ExecuteComputation with the RunPostOrder and FetchTensorData, however, that did not help. The two methods seems to run only after the previous execution is completed. This could be because of the [device lock present in CollectSyncTensors](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1180) (not completely sure though).

## To Reproduce

1. Apply the below patch on [MNIST training script](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py) . The patch is mainly to add logic to simulate gradient accumulation and add profiling.
```
From d05f582cdf0b8ec97399c04fdcd1769bc44b2821 Mon Sep 17 00:00:00 2001
From: EC2 Default User <ec2-user@ip-172-31-36-175.ec2.internal>
Date: Mon, 21 Mar 2022 21:29:29 +0000
Subject: [PATCH] profiling

---
 test/test_train_mp_mnist.py | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/test/test_train_mp_mnist.py b/test/test_train_mp_mnist.py
index f4a2e8e4..ecf09309 100644
--- a/test/test_train_mp_mnist.py
+++ b/test/test_train_mp_mnist.py
@@ -24,7 +24,7 @@ import torch_xla.utils.utils as xu
 import torch_xla.core.xla_model as xm
 import torch_xla.distributed.xla_multiprocessing as xmp
 import torch_xla.test.test_utils as test_utils
-
+import torch_xla.debug.profiler as xp
 
 class MNIST(nn.Module):
 
@@ -112,6 +112,7 @@ def train_mnist(flags, **kwargs):
   lr = flags.lr * xm.xrt_world_size()
 
   device = xm.xla_device()
+  server = xp.start_server(9014)
   model = MNIST().to(device)
   writer = None
   if xm.is_master_ordinal():
@@ -123,13 +124,14 @@ def train_mnist(flags, **kwargs):
     tracker = xm.RateTracker()
     model.train()
     for step, (data, target) in enumerate(loader):
-      optimizer.zero_grad()
-      output = model(data)
-      loss = loss_fn(output, target)
-      loss.backward()
-      xm.optimizer_step(optimizer)
-      tracker.add(flags.batch_size)
+      with xp.Trace('model_build'):
+        optimizer.zero_grad()
+        output = model(data)
+        loss = loss_fn(output, target)
+        loss.backward()
       if step % flags.log_steps == 0:
+        xm.optimizer_step(optimizer)
+        tracker.add(flags.batch_size)
         xm.add_step_closure(
             _train_update,
             args=(device, step, loss, tracker, writer),
@@ -186,4 +188,4 @@ def _mp_fn(index, flags):
 
 
 if __name__ == '__main__':
-  xmp.spawn(_mp_fn, args=(FLAGS,), nprocs=FLAGS.num_cores)
+  xmp.spawn(_mp_fn, args=(FLAGS,), nprocs=FLAGS.num_cores)
\ No newline at end of file
-- 
2.32.0


```

2. Run the training script for more number of epochs and with high number of workers. In my case, I ran it for 8
3. Start the tensorboard in a separate terminal and record the profile. Since each training step takes very small amount of time, I would recommend running the profile only for like 100-200ms. This should be enough to record the bottleneck.

## Expected behavior

Ideally, we would expect that this time is negligible. If we can avoid this 7% overhead from each execution that can help lower the overall training time.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: GPU
 - torch_xla version: 0.10

## Additional context
A similar behavior is observed for BERT-large too. Even there we observe around 5-7% of the time spent between two executions.

**Comment by JackCaoG** [comment]
[RunPostOrder](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1272) will run a post order traversal on all live tensors and provide a po_data that will be used for lowering_context to lower to HLO text. It also walk through all tensors and figure out which tensor would be the parameter(input) of the graph. Even for a cached graph this step is necessary since we need to figure out the input to the cache graph.

[FetchTensorData](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1312) on the other hand 
determine where the output of the computation should go, and we will update tensor value after the execution in [here](https://github.com/pytorch/xla/blob/90af49ca836c5fe34fc80d524dc1b3cc6cbdb93b/torch_xla/csrc/tensor.cpp#L1368).

These two steps will be blocked on previous execution finish since we need to wait until last execution to update all live tensor's value. For example

```
a = torch.tensor(1)
b = torch.tensor(2)
c = a + b
xm.mark_step()
d = c * 2
xm.mark_step()
```

In second mark_step we don't want to re-execute `c` hence we need to wait for previous execution and update to happen. Then in second mark_step we see `c` become a device_data and it become the input to the second graph.

**Comment by JackCaoG** [comment]
@wconstab and his team is exploring how to optimize the tracing if the graph being traced is identical. Maybe there is a way to avoid traversing the graph to determine the `data` for input and output of the graph?

**Comment by wconstab** [comment]
Tagging @desertfire as he is taking a look at tracing performance

**Comment by ymwangg** [comment]
One interesting and promising solution to avoid this overhead is to introduce graph `record` and `replay` API similar to the pytorch cuda graph feature (https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/), so that the user has more control over the graph tracing + compilation + execution pipeline.

**Comment by aws-rhsoln** [comment]
For short-term, is there a way we can optimize those two methods? Also, do we have any timeline on the tracing work?

**Comment by desertfire** [comment]
> For short-term, is there a way we can optimize those two methods? Also, do we have any timeline on the tracing work?

The problem is really about the synchronization from `DeviceLockWait`, not that the two functions themselves are computation intensive. I am trying to see if there is a way to get rid of that for GPU on the lazy_tensor_staging branch. Just confirm, your case is also running on GPU, right?

For the tracing speed improvement work, we are trying to have something more concrete in the coming weeks.

**Comment by aws-rhsoln** [comment]
In this example, yes its running on GPU. Is your work specific only to GPU or is it generic enough to extend for other devices too?

**Comment by desertfire** [comment]
@aws-rhsoln , can you cherry-pick https://github.com/pytorch/pytorch/pull/74864 for a try? It is a change on the lazy_tensor_staging branch, but its logic applies to the pytorch/xla code as well.

**Comment by aws-rhsoln** [comment]
Cool, I can try that. Will update the results with the change. Thanks!

**Comment by JackCaoG** [comment]
@aws-rhsoln I tried https://github.com/pytorch/xla/pull/3457 and it failed some test.

**Comment by aws-rhsoln** [comment]
Oh okay.. Will monitor the PR and once you have a fix, can try it out. Thanks for pointing out.
